File: leetcode/C++/convert-sorted-list-to-binary-search-tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> v;
    TreeNode* traverse(int l, int r){
        if(l>r) return nullptr;
        int m = l+(r-l)/2;
        return new TreeNode(v[m],traverse(l,m-1),traverse(m+1,r));
    }
    TreeNode* sortedListToBST(ListNode* head) {
        for(ListNode* cur=head; cur!=nullptr; cur=cur->next) v.push_back(cur->val);
        if(v.empty()) return nullptr;
        return traverse(0,v.size()-1);
    }
};

File: leetcode/Python/convert-sorted-list-to-binary-search-tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.v = list()
    def traverse(self,l,r):
        if l>r:
            return None
        m = l + (r-l)//2
        return TreeNode(self.v[m],self.traverse(l,m-1),self.traverse(m+1,r))
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        while head is not None:
            self.v.append(head.val)
            head=head.next
        if len(self.v) == 0:
            return None
        return self.traverse(0,len(self.v)-1)